File: sol/solution/0001-0100/0025. Reverse Nodes in k-Group/Solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func reverseKGroup(_ head: ListNode?, _ k: Int) -> ListNode? {
        var curr = head
        var count = 0
        while curr != nil && count < k {
            curr = curr?.next
            count += 1
        }
        if count == k {
            let reversedHead = reverseLinkedList(head, k)
            head?.next = reverseKGroup(curr, k)
            return reversedHead
        }
        return head
    }
    private func reverseLinkedList(_ head: ListNode?, _ k: Int) -> ListNode? {
        var prev: ListNode? = nil
        var curr = head
        var remainingK = k
        while remainingK > 0 && curr != nil {
            let next = curr?.next
            curr?.next = prev
            prev = curr
            curr = next
            remainingK -= 1
        }
        return prev
    }
}

File: sol/solution/0001-0100/0025. Reverse Nodes in k-Group/Solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let k_usize = k as usize;
        let mut count = 0;
        let mut ptr = &head;
        while let Some(node) = ptr {
            count += 1;
            if count == k_usize { break; }
            ptr = &node.next;
        }
        if count < k_usize {
            return head;
        }
        let (mut reversed_head, remainder) = Self::reverse_linked_list(head, k_usize);
        if let Some(ref mut node) = reversed_head {
            let mut tail = node.as_mut();
            while let Some(ref mut next_node) = tail.next {
                tail = next_node.as_mut();
            }
            tail.next = Self::reverse_k_group(remainder, k);
        }
        reversed_head
    }
    fn reverse_linked_list(mut head: Option<Box<ListNode>>, k: usize) -> (Option<Box<ListNode>>, Option<Box<ListNode>>) {
        let mut prev: Option<Box<ListNode>> = None;
        let mut curr = head.take();
        let mut remaining_k = k;
        while remaining_k > 0 {
            if let Some(mut node) = curr {
                let next = node.next.take();
                node.next = prev;
                prev = Some(node);
                curr = next;
                remaining_k -= 1;
            } else {
                break;
            }
        }
        (prev, curr)
    }
}